### ISSUE TITLE
Changes to package-lock.json trigger an Affected on all packages

### ISSUE THREAD

**Issue by pedrolourencoribeiro** [nrwl/nx#15116]
### Current Behavior

Update one dependency that affects only one package
run ` npx nx print-affected --select=projects --base=main` 
will print all packages

### Expected Behavior

Update one dependency that affects only one package
run ` npx nx print-affected --select=projects --base=main` 
should print only the affected package

### GitHub Repo

https://github.com/pedrolourencoribeiro/nx-affected

### Steps to Reproduce

1. clone git repo https://github.com/pedrolourencoribeiro/nx-affected
2. checkout zod-update
3. install dependencies (`npm i`)
4. run ` npx nx print-affected --select=projects --base=main`

### Nx Report

```shell
>  NX   Report complete - copy this into the issue template

   Node : 16.18.0
   OS   : darwin arm64
   npm  : 8.19.2
   
   nx                      : 15.7.2
   @nrwl/linter            : 15.7.2
   @nrwl/workspace         : 15.7.2
   @nrwl/cli               : 15.7.2
   @nrwl/devkit            : 15.7.2
   @nrwl/eslint-plugin-nx  : 15.7.2
   @nrwl/js                : 15.7.2
   @nrwl/tao               : 15.7.2
   typescript              : 4.8.4
```

### Additional Information

This was changed in the version 15.4.0 => https://github.com/nrwl/nx/pull/13960#discussion_r1055559215
I don't fully know the reason but this is blocking us from upgrading to a new version since we have quite a lot of packages and we want to know what is really affected when updating a dependency.
Is there a workaround for this? From looking at the code nothing came to my mind but maybe there is.

Thank you for all the work btw the project is amazing and its saving us a lot of time 🙇‍♂️
Ps: I'm available to contribute to this if needed :)

**Comment by ianldgs** [comment]
Hi @AgentEnder / @meeroslav. Sorry to bother again about this, but since we like to release affected projects often, this issue causes our CI to be very busy whenever someone adds a dependency to a local package.json.

Would you consider accepting the changes of https://github.com/nrwl/nx/pull/15776 under a feature flag? (Until a final solution is implemented).
That way, anyone can opt out of this behaviour at their own risk.

**Comment by brandoncollins7** [comment]
I am having the same issue, this is a big problem:(

**Comment by meeroslav** [comment]
Hey all,

The logic behind this is not as trivial as you would have expected.

## Explanation

The reason why `print-affected` and the `affected` itself mark all the projects is because every change in the lock file **can** affect all of them. You run `nx` commands via executors (e.g. `nx:run-commands` or `@nrwl/js:tsc`). Without knowing what target (and what executor) you are running it's impossible to tell whether the project is affected or not so it's safer to mark it as (potentially) affected.

Imagine you have a `react` app, and you update `rxjs`. Although your project might not depend on it, you might have an executor that looks like this:

```typescript
import { range } from 'rxjs';

export default async function buildExecutor(
  options: Schema,
  context: ExecutorContext
) {
  // ...your executor code which uses the range from rxjs
}
```
Now each build of your app is actually affected by changes in the `rxjs`.

## Caching to the rescue

Despite every project is marked as affected, once you run the actual command (or task) we will have more information and can at that point decide if the project had actually changed or not. The task consists of the following parameters:
- sources of the project (_should be intact_)
- command (_unchanged_)
- target executor's dependencies (e.g. `@nrwl/js` and everything it depends on)

If your target hasn't been affected by the change, you will get instant replay from the cache, otherwise as long as any of those is different we will run the full command.

### Known executors only

This unfortunately only works with known executors (e.g. `@nrwl/*`) where we can guarantee the list of dependencies. For all the other executors, including the `nx:run-commands` we will fall back to **all affected** mode since we have no idea what that executor might be running and which packages it depends on.

## Workaround

There is a workaround that will land in v16 where you can skip checking `package.json` and lock files.
In your `nx.json` you need to add the following:
```json
{
  "pluginsConfig": {
    "@nrwl/js": { "analyzePackageJson": false }
  }
}
```

This will disable all the checks so your graph will no longer have any of the external packages included as a node.

**Comment by meeroslav** [comment]
As you can see @pedrolourencoribeiro once we built the packages on `main`, then switch to `zod-update` and run the same command, not only was `is-odd` not re-run (we got results back from cache) but also `is-even` was not re-run.

Although `package.json` and `package-lock.json` were changed, the actual installed version of `zod` was not changed, so even `is-even` was not affected (no pun intended).

Unless you are using the result of `print-affected` for some custom commands or using custom executors, having all projects affected with lock file changes doesn't necessarily mean all of these would be actually run.

![Screenshot 2023-04-20 at 14 32 04](https://user-images.githubusercontent.com/881612/233366724-c30829fa-2877-489c-a496-97a4446b1903.png)

**Comment by meeroslav** [comment]
If you believe our docs are missing this information, please open another issue where you would specify that and feel free to link it to this one.

In the meantime, I will close this one.

**Comment by ianldgs** [comment]
@meeroslav would changing a project's package.json (pnpm workspaces) still mark the package as affected?
And it's there an option to set this option globally?

**Comment by meeroslav** [comment]
@ianldgs, changing project's `package.json` should mark that project as affected, since `package.json` is a file belonging to that project.

**Comment by meeroslav** [comment]
@kuba-gaj, the `rxjs` might not be specified in any of the `package.json`s, if it's just a transitive dependency of a given executor (let's call it `@myorg/awesome-executor`). 

So, if my root `package.json` doesn't have a dependency to `rxjs` explicitly, but one of the packages (let's call it `packageA`) does, it's very likely that `@myorg/awesome-executor` would share the same dependency.

So if I bump the version of `rxjs` in my `packageA`, that would install a new version or `rxjs`, update the lock file, and potentially influence the `@myorg/awesome-executor` (as long as range matches).

Yes, toggling `analyzePackageJson` is a big hammer that might make more damage than help solves the issue.

If I may ask, what is it that currently doesn't work for you? Having all projects marked as `affected` is generally not an issue, as long as caching properly picks it up and skips running the task.

Can you explain what commands you are running and how is the current implementation causing you trouble?

Based on my experience, the packages are rarely updated and when they are caching can save a lot of work if changed packages do not affect a given project.

**Comment by meeroslav** [comment]
If I understand correctly, you are doing the following:

- Change a package, which makes it affected
- An automated release kicks in, publishes that package, and changes the version of that package's `package.json`
- (Depending on the package manager) this might cause a change in the root lock file
- This would mark all packages as affected

What happens next? What does your targets look like?

Was caching able to skip all the other packages since they were not affected?
Is that package being released again with bumped version because it was affected by the change in `package.json`?

**Comment by ianldgs** [comment]
Our pipeline looks roughly like this:

- `nx affected --target=version`
  - https://github.com/jscutlery/semver
  - on PRs, only bumps versions in the package.json (e.g. 1.0.0 to 1.0.1)
  - on merge to main, bumps versions, creates a commit and tag (e.g. 1.0.0 to 1.0.1 again, so that it gets cache from the PR run)
  - we found it was safer to invalidate cache on PRs instead of on main
- `nx affected --targets=lint,test,build,docker`
  - because of the version bump, cache is invalidated any lib has to re-run lint, test and build
  - apps have to run docker (https://github.com/gperdomor/nx-tools/tree/main/packages/nx-container)
- `nx affected --target=deploy`
  - https://github.com/edbzn/ngx-deploy-npm for libs
  - on PRs, it's just a `--dryRun`
  - on merge to main, publish to our internal registry
  - it makes more sense for us to do like this because we don't have a monorepo, but a shared repo where libs live
  - other apps in other repos have to install them from the registry, and once a fix that got merged, we immediately want it available to install in the other repo
- `nx affected --target=upgrade`
  - proprietary upgrade script for *some* (not all) applications
  - on merge to main, upgrades the applications in production

So when one lib or app adds one dependency, say via `pnpm install --filter {my-lib} {some-dependency}`, the lockfile makes everything affected, triggering a release cycle for all projects in the repo, and a full rebuild.

**Comment by kuba-gaj** [comment]
We have an existing monorepo with ~60 services and ~60 packages that we added nx to. 
Some operations do not use caching:
- `typecheck` - we separated typechecking to another job using npm script `tsc --noEmit` and use `esbuild` during `build` step. AFAIK there is no official executor to perform typechecking only, so even though we have caching enabled for this step, it won't work
- `deploy` jobs (for each environment) - this is not cachable, it deploys different projects using serverless framework / ecs / eks. Updating a single dependency now causes 60 services to deploy
- each ecs deploy job creates an artefact that is later used to confirm successful deployment in ecs in the following `wait_for_deployment` step that queries the status of deployments for each service until it's in a stable state
- we have e2e tests that we want to run only if a subset of services is affected, for example, any of 6 services that are tested by that workflow

**Comment by kuba-gaj** [comment]
@meeroslav, and just to come back to your example with `rx`. Is that a problem only with `transitive` dependencies in executors? Because currently (with `yarn3`) we have a separate `package_validation` script on CI that doesn't allow to `import/require` packages that are not specified in the project's `package.json` as a dependency.

We are only considering a move to `pnpm`, and as I understand, at least with the default configuration and hoisting disabled, it would also be impossible to require `rx`.

From their docs here: https://pnpm.io/symlinked-node-modules-structure

```
A great bonus of this layout is that only packages that are really in the dependencies are accessible. With a flattened node_modules structure, all hoisted packages are accessible.
```

**Comment by brandoncollins7** [comment]
I am doing same for pnpm-lock.yaml.... I have been getting some off dependency related issues I don't know if it's related.

**Comment by ianldgs** [comment]
What kind of issues @brandoncollins7?
And genius solution @smailc! Haven't thought of it. Will try.

**Comment by kuba-gaj** [comment]
@meeroslav thank you for your answer. Unfortunately, we have a lot of tasks still using generic executors, so caching is not working at the moment. We will have to experiment more, maybe try ignoring the lockfile as mentioned above and monitoring risks for now. We also want to introduce renovateBot that will further increase number of changes to global lockfile

**Comment by meeroslav** [comment]
I am happy to tell you we are actually working on a solution for this. It's currently in the design phase but you should expect it in a few weeks.

In the meantime, you can use `.nxignore` as @smailc suggested.

**Comment by meeroslav** [comment]
I'm still keeping this closed as the affected works as expected, but caching is currently a bit limiting (which is why we are making changes to it)

**Comment by brandoncollins7** [comment]
One reason caching does not work is because our pipeline relies on building a docker image. You can't cache a docker image and so when nx affected is triggered on all projects it forces us to rebuild all docker containers

**Comment by meeroslav** [comment]
You can (and should) cache image creation and build. This is the recommended way.

What you cannot cache is push.

**Comment by brandoncollins7** [comment]
Like docker cache? Or are you talking about something else? Do you have any examples? I don't understand how nx could cache a large docker build?

**Comment by meeroslav** [comment]
You can check 
https://www.npmjs.com/package/@nx-tools/nx-container and 
https://nx.dev/node-server-tutorial/5-docker-target for more information.

We are preparing a blog post on this topic soon.

**Comment by brandoncollins7** [comment]
But even still, let's say I use docker cache. I still have to build and push all new containers as my development environment builds and pushes containers tagged with their git hash. That means not only to I have to build and push, but I also have all of my kubernetes pods restart and reload a new container even though nothing in the apps have changed.

**Comment by meeroslav** [comment]
I understand your pain and can inform you that solution is on the way.

We are implementing a logic to restrict hash scope for custom targets.

This would for example allow you to specify that your docker push target is not affected by any changes in the installed packages and solely depends on the project's built artifact.

**Comment by bojanbass** [comment]
Nice conversation guys!

I'm also surprised it is possible to set docker build or deploy tasks as cacheable, it's quite strange to me, as there are no artifacts in the dist folder created. 

I also have another case where we have npm and python based apps inside the same monorepo. I would really like to exclude py apps from being affected after changing package.json, as there's no connection between those. Is this possible to implement?

Thank you.

**Comment by bojanbass** [comment]
@meeroslav, just a side question, partially connected to this issue. If each project has its own `deploy` folder and I would like to run `helm lint` task only for affected ones, is it possible to detect only if a part of the project was affected? E.g. if I change deploy folder, I would run `helm-lint` task, but I the src was changed, I would run `eslint`?

**Comment by brandoncollins7** [comment]
> I understand your pain and can inform you that solution is on the way.
> 
> We are implementing a logic to restrict hash scope for custom targets.
> 
> This would for example allow you to specify that your docker push target is not affected by any changes in the installed packages and solely depends on the project's built artifact.

Not sure if this will work, as unfortunately for my build process using solely affected doesn't work. I use the output from print-affected and then have a custom process around the output of that. If only the container task is ignored then print-affected would still output all projects?

**Comment by brandoncollins7** [comment]
> I understand your pain and can inform you that solution is on the way.
> 
> We are implementing a logic to restrict hash scope for custom targets.
> 
 Any updates?

**Comment by neon-inkblast** [comment]
Also keen to see something that will work for this with pnpm, as any single project changing dependencies propogates to the pnpm-lock and currently marking all of our projects as affected.

**Comment by meeroslav** [comment]
@bojanbass you can use named inputs to mark what file sets influence which targets - read more about it here https://nx.dev/more-concepts/customizing-inputs and here https://nx.dev/reference/project-configuration#inputs-&-namedinputs

**Comment by meeroslav** [comment]
@aequasi you can now use the new `externalDependencies` target input to restrict those situations

**Comment by meeroslav** [comment]
@brandoncollins7 there is a difference between affected projects, affected targets, and runtime cache hits. I think there's an issue in understanding the differences between those three.

If you change the code in one of the projects, only that project and its dependents will be affected. There is no way how anything else might be affected. Note that when we say affected, what we actually mean is "potentially affected".

If you change the lock file (bump version of any package) all projects are affected (depending on the executor they may or may not be actually affected but at this point, we can't tell). If you specify the `target` (like you would for `print-affected`) you are simply limiting the list of projects to those that contain such target e.g. `deploy`.

We can only tell if a project is actually affected by running the full task hashing and calculating the exact dependencies. Task hash depends also on runtime conditions (node version, environment variables, the exact command being run with all the flags, etc.) and those we cannot possibly know unless the command has been actually run. This was the reason we introduced the `externalDependencies` input to give you the power to control this. 

If you are using one of the `@nx/*` plugin executors for running your regression tests, you would have already benefitted from surgical external deps sensitivity - only dependencies of your project (plus its dependencies), and the given executor would trigger the cache miss. For third-party executors and standard commands, you need to use `externalDependencies` to restrict the packages that are included in the hash.

If you are using `print-affected` to get the list of projects and then do a calculation on your own, then you are losing all the caching and hashing benefits. Your deploy can be a cacheable operation, and instead of reimplementing this on your own, you should use the built-in functionality where nx will run deploy only when things have been changed.

**Comment by meeroslav** [comment]
We will create documentation explaining this in more detail, as it's obviously not clear enough.

cc @juristr

**Comment by brandoncollins7** [comment]
I do not understand how building and pushing a docker container stamped with the version number (git hash) and updating kubernetes can be a cachable operation.

**Comment by brandoncollins7** [comment]
We are using nx affected for our development environment, but because of this issue we have been forced to implement changesets as a build trigger instead of using nx affected. We only build and deploy projects which have a changeset associated with them for our staging and production environments.

**Comment by meeroslav** [comment]
@brandoncollins7,

Your deployment can have several steps:
- Building a docker image
- Pushing image to k8s

You can control both of these steps with inputs, but the suggested process would be:
- Have a docker image building cacheable that is only triggered by build and has `externalDependencies: []` to make sure no changes in the lock file trigger it directly. This caches the docker image
- Your push is marked as a cacheable operation that again only depends on the `docker` target changes and has `externalDependencies: []`

This means that unless the build was triggered (and not taken from the cache), your `docker` and `deploy` steps would not do anything.

Does that make sense?

**Comment by brandoncollins7** [comment]
Hmm, I'm not sure I understand how to make sure docker and deploy are only triggered by build? Currently I have my docker task dependsOn["build"], is there something else that needs to be done? Currently my docker task is not cachable since I didn't think it's outputs could be cached.

Also my deploy is 3 steps:
1. Build docker image and tag with current git hash (for development) or current changeset version (for staging/production)
2. Push image to aws image store
3. Update argocd k8's deployment with git hash or version number

I will look more into what you've said however, see if I can make something work...

**Comment by brandoncollins7** [comment]
I'm also a bit confused on how to use externalDependencies. Do I have to manually list every npm package each project uses in order to take advantage of this and actually build if a dependency that belongs to that project changes?

**Comment by ianldgs** [comment]
It'd be great to have an article/tutorial with examples of a docker ([nx-container](https://github.com/gperdomor/nx-tools/tree/main/packages/nx-container)) or npm ([semver](https://github.com/jscutlery/semver) & [ngx-deploy-npm](https://github.com/bikecoders/ngx-deploy-npm)) target that uses the `externalDependencies` option.

It's usually docker push or npm publishing that will suffer when the lockfile affects every project.

**Comment by brandoncollins7** [comment]
How do I do this in a nx distributed build? If first I run my lint and build task.... then when I go to run my container and deploy tasks, which dependOn build, they will see the build has already been cached and so they don't deploy.

**Comment by brandoncollins7** [comment]
So I have been working long hours trying to implement this in my CI and everything seemed to be going great until this:

![image](https://github.com/nrwl/nx/assets/36413482/d39385d4-3fa0-4dec-9cdc-0a9dc3c3ab3b)

I have no idea why my deploy task got a cache miss, which effectively told kubernetes to deploy pod images that were never pushed. According to this compare screen it should have been a cache hit. I am confused.

**Comment by rv2673** [comment]
> We will create documentation explaining this in more detail, as it's obviously not clear enough.

@meeroslav do you have update on that eics. It would be incredibly helpful I think.

> If you are using `print-affected` to get the list of projects and then do a calculation on your own, then you are losing all the caching and hashing benefits. 

@meeroslav what woul be a way to hook into that calculation if print-affected skips it.

To give a bit of context. We have 3 artifacts that we want to deploy.

If the first artifact has changed/been built no cache hit we want to deploy second artifact with different parameters. The same goes for the third depending on the second artifact.

The deploys however need to run in restricted environment based on artifacts. And there is a queue mechanism that we would like to avoid triggering if nothing has changed, which we can only affect by setting certain variables.

So i would like to set 3 variables with value whether any of the 3 artifacts had been (really/not potentially) affected and was not loaded from cache.

I thought to use print-affected but that only shows potentially affected. So I am kindof stuck. What is the recommend nx way to hook something in on the full calculation, would this task have run without actually running the task and caching the result?

**Comment by meeroslav** [comment]
> do you have update on that docs

Not yet

> what would be a way to hook into that calculation

Currently, there is no way to extract that information as far as I know. The calculation happens in realtime as you execute the target. But maybe @FrozenPandaz would know more.

Btw. the `print-affected` is deprecated. You should use [nx show](https://nx.dev/packages/nx/documents/show)

**Comment by BioCarmen** [comment]
> > @brandoncollins7,
> > Your deployment can have several steps:
> > 
> > * Building a docker image
> > * Pushing image to k8s
> > 
> > You can control both of these steps with inputs, but the suggested process would be:
> > 
> > * Have a docker image building cacheable that is only triggered by build and has `externalDependencies: []` to make sure no changes in the lock file trigger it directly. This caches the docker image
> > * Your push is marked as a cacheable operation that again only depends on the `docker` target changes and has `externalDependencies: []`
> > 
> > This means that unless the build was triggered (and not taken from the cache), your `docker` and `deploy` targets would not do anything.
> > Does that make sense?
> 
> So I have been working long hours trying to implement this in my CI and everything seemed to be going great until this:
> 
> ![image](https://user-images.githubusercontent.com/36413482/254608866-422254ac-d641-44ca-ade9-7cb289406309.png)
> 
> I have no idea why my deploy task got a cache miss, which effectively told Kubernetes to deploy pod images that were never pushed. According to this compare screen it should have been a cache hit. I am confused and stuck with this pretty picture:( ![image](https://user-images.githubusercontent.com/36413482/254609225-ecd3bd06-80ea-457c-9df6-49c2f567481a.png)
> 
> I don't know if it could have something to do with the fact that I ran the final deploy task (which depends on the cached tasks) outside of distributed execution?
> 
> ```
> - name: Run commands in parallel
>         env:
>           INPUT_GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
>         run: |
>           pids=()
> 
>           # list of commands to be run on agents
>           VERSION=${{ github.sha }} ENVIRONMENT=uat npx nx affected --base=\$NX_BASE --head=\$NX_HEAD -t container:ci --parallel=3 --nx-bail=true --configuration=uat &
>           pids+=($!)
>           echo "PID for nx container: ${pids[-1]}"
> 
>           # run all commands in parallel and bail if one of them fails
>           for pid in "${pids[@]}"; do
>             echo "Waiting for pid $pid"
>             if ! wait $pid; then
>               echo "Process $pid exited with status $?"
>               exit 1
>             fi
>           done
> 
>           exit 0
> 
>       - name: Stop all running agents for this CI run
>         # It's important that we always run this step, otherwise in the case of any failures in preceding non-Nx steps, the agents will keep running and waste billable minutes
>         if: ${{ always() }}
>         run: npx nx-cloud stop-all-agents
> 
>       - name: Update GitOps
>         run: |
>           git pull
>           NX_CLOUD_DISTRIBUTED_EXECUTION=false VERSION=${{ github.sha }} ENVIRONMENT=uat npx nx affected --base=\$NX_BASE --head=\$NX_HEAD -t deploy --parallel=3 --nx-bail=true --configuration=uat
> ```
> 
> But I have to run it outside of distributed execution because all the deploy task does is write the image tags to some k8 files in the repo and checks them in. If it's distributed I will multiple conflicting git commits to make across machines which is likely to result in git errors.

Hello @brandoncollins7 did you get it to work? facing the same issue
